util/shlib_wrap.sh.in

```diff
@@ -30,6 +30,13 @@ fi
 THERE="`echo $0 | sed -e 's|[^/]*$||' 2>/dev/null`.."
 [ -d "${THERE}" ] || exec "$@"	# should never happen...
 
+if [ -z "$OPENSSL_ENGINES" ] && [ -d "${THERE}/engines" ] ; then
+	OPENSSL_ENGINES="${THERE}/engines"; export OPENSSL_ENGINES
+fi
+if [ -z "$OPENSSL_MODULES" ] && [ -d "${THERE}/providers" ] ; then
+	OPENSSL_MODULES="${THERE}/providers"; export OPENSSL_MODULES
+fi
```

Comment:
I would rather not at this level.  What if I want to test a newly built test program against the default config file?

Reply:
Hmmm... too bad. It was such a convenient way to plug in the debugger into a test command. Do you have a better suggestion how to fix it?

Reply:
The variables could be set explicitly on the command line instead of implicitly by the _test target, but that is not really a satisfying solution either.

Reply:
I don't see much inconvenience with this

```
OPENSSL_MODULES=providers ./util/shlib_wrap.sh gdb test/evp_extra_test
```

Reply:
Often it's more like

```
OPENSSL_MODULES=../../providers ../../util/shlib_wrap.sh gdb ../../apps/openssl  ...
```

> I don't see much inconvenience with this

You mean inconvenience to add it manually, or inconvenience to add it to the test output?

Reply:
Inconvenience to add manually.
As for the test output, please remember that it's not Unix only

Reply:
Well, then it seems like it's not worth pursue this pull request any further...

Reply:
> I would rather not at this level. What if I want to test a newly built test program against the default config file?

I don't get the point. Why would setting `OPENSSL_MODULES` and `OPENSSL_ENGINES` prevent me from doing it?

Reply:
> > I would rather not at this level. What if I want to test a newly built test program against the default config file?
> 
> I don't get the point. Why would setting `OPENSSL_MODULES` and `OPENSSL_ENGINES` prevent me from doing it?

ping @levitte

Reply:
Ah, sorry. There us also a built-in modules and engines location, which are only accessible when `OPENSSL_MODULES` and `OPENSSL_ENGINES` are unset, resp.
So please explain to me how to unset those variables when this script sets them.

Reply:
(the logic is exactly the same with the default config file, that's why I mixed them up)

Reply:
> Ah, sorry. There us also a built-in modules and engines location, which are only accessible when `OPENSSL_MODULES` and `OPENSSL_ENGINES` are unset, resp.
> So please explain to me how to unset those variables when this script sets them.

The built-in location is determined at configuration time and it determines where the modules [and engines] will end up after installation, right?

My understanding is that the two scripts shlib_wrap.sh and opensslwrap.sh are meant to test openssl after building and before installing it. This means that in  most cases it is desired to have those two environment variables set pointing into the build tree, whereas the other case where you want to load the modules from the install location is exceptional.

You are right, currently it is not possible to unset the pathes explicitly. What if I would add a mechanism to unset them, for example like this? Would my change be then acceptable for you?

    OPENSSL_MODULES=unset  [OPENSSL_ENGINES=unset]  ../util/shlib_wrap.sh apps/openssl ....

Reply:
No. Come on, you're adding extra complexity to fix something that isn't a problem in the first place.

Reply:
Basically, there is a reason why there are two scripts. How about leaving the comfort stuff with opensslwrap? If you want a similar wrapper for test programs, then add one, but please let shlib_wrap do what it's name implies and just that.

Reply:
> Basically, there is a reason why there are two scripts. How about leaving the comfort stuff with opensslwrap? 

Correct. shlib_wrap.sh is a generic helper to run and test openssl programs in the source tree without installing them. And opensslwrap.sh is just a simple comfort addon on top of it which nobody really needs. I don't use opensslwrap.sh at all, because it only wraps the openssl app and nothing else.

But shlib_wrap.sh is much more flexible and was always a great help when pluggin in gdb into failed commands from the testsuite. It just worked.

Sigh. Am I the only @openssl team member who mourns about the loss of the ability to just grab the test output and run it in the debugger? 

> If you want a similar wrapper for test programs, then add one, but please let shlib_wrap do what its name implies and just that.

If I'm the only one, then I'll stop complaining and just write my own wrapper which I place into my own `~/bin`. But I certainly won't create pull requests for useless addons which nobody seems to need.

Reply:
> but please let shlib_wrap do what its name implies and just that.

Isn't that exactly it's purpose: to ensure that the libraries from the build tree are loaded and not the installed ones?

Reply:
I use it for GDB sessions.  It would be bad to lose this capability.

Reply:
So does this mean you are pro or contra my change that shlib_wrap.sh sets OPENSSL_MODULES and OPENSSL_ENGINES automatically to the path in the build tree @paulidale?

Reply:
> ...sets OPENSSL_MODULES and OPENSSL_ENGINES automatically...

...if unset...

Reply:
I view plug-ins (which our providers and engines are) as something different than shared libraries, *even if the technology is the same*. The difference is build-time linking vs run-time linking.

Reply:
I have an alternative idea that you might suitable, if you can have a little bit of patience? PR coming up.

Reply:
Sure. Surprise me... ;-)

Reply:
Have a look at #11110, I believe it fulfils what you want.